### PR TITLE
[XLA:GPU] HorizontalLoopFusion now generates a cleaner graph

### PR DESCRIPTION
**Pull request by nouiz**
This ask less works by other optimization pass.
This is a safe subset of another PR. 
In case of revert, it will revert less changes.
The main changes are:
- Only insert bitcast when needed.
- Only insert reshape when needed.
- Remove old HloComputation when not needed.

This also update the tests for the above change.

### CODE REVIEW

File: tensorflow/compiler/xla/service/copy_insertion_test.cc
Comment (cheshire) [inline tensorflow/compiler/xla/service/copy_insertion_test.cc:2467]: Please move to a separate PR

File: tensorflow/compiler/xla/service/gpu/horizontal_loop_fusion.cc
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/horizontal_loop_fusion.cc:405]: I understand this change does not insert a reshape unless necessary, is this the "main" change associated with the PR? If yes, could it be reflected in PR description?

File: tensorflow/compiler/xla/service/gpu/horizontal_loop_fusion.cc
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/horizontal_loop_fusion.cc:487]: Was the computation previously DCEd?

**Review by cheshire — CHANGES_REQUESTED** [review]
Thanks! Generally good to go modulo minor nit, commit could be more self-descriptive

**Comment by gbaned** [comment]
Hi @nouiz Can you please check @cheshire's comments and keep us posted ? Thank you!

File: tensorflow/compiler/xla/service/gpu/horizontal_loop_fusion.cc
Comment (nouiz) [reply tensorflow/compiler/xla/service/gpu/horizontal_loop_fusion.cc:487]: Yes, here:
https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/compiler/xla/service/gpu/gpu_compiler.cc;l=600

Note, that DCE is still needed. But it will do less work.

File: tensorflow/compiler/xla/service/copy_insertion_test.cc
Comment (nouiz) [reply tensorflow/compiler/xla/service/copy_insertion_test.cc]: Done: https://github.com/tensorflow/tensorflow/pull/57042

File: tensorflow/compiler/xla/service/gpu/horizontal_loop_fusion.cc
Comment (nouiz) [reply tensorflow/compiler/xla/service/gpu/horizontal_loop_fusion.cc:405]: I updated the description with the 3 main changes. This is one of them.

**Review by nouiz — COMMENTED** [review]
I did all the requested changes.

File: tensorflow/compiler/xla/service/gpu/horizontal_loop_fusion.cc
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/horizontal_loop_fusion.cc:470]: Nit: rephrase? Does this run after algebraic simplifer which should clean bitcasts up?

File: tensorflow/compiler/xla/service/gpu/horizontal_loop_fusion.cc
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/horizontal_loop_fusion.cc:475]: They are not "bitcasts" anymore?

File: tensorflow/compiler/xla/service/gpu/horizontal_loop_fusion.h
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/horizontal_loop_fusion.h:97]: "only if needed for shapes to match" or similar? Because I understand reshapes are added not for being "useful", but because the shapes don't agree without them?

File: tensorflow/compiler/xla/service/gpu/horizontal_loop_fusion.cc
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/horizontal_loop_fusion.cc:403]: Similarly to below, these are not "reshapes" anymore?

**Review by cheshire — CHANGES_REQUESTED** [review]
Sorry few more nits. Could we also simplify the commit description? The first paragraph and the last sentence do not seem to add information, maybe simply: "simplify resulting HLO from horizontal fusion: (cases description)?

**Comment by nouiz** [comment]
I fixed all the comments in their respective commits and updated each commit description.

**Comment by cheshire** [comment]
> I fixed all the comments in their respective commits and updated each commit description.

When the system imports the PR, it squashes all the commits into one. It squashes the commit descriptions as well. This usually results in very confusing commits messages. Would be best to have a single commit, with a single descriptive message.

**Comment by nouiz** [comment]
I manually squashed the commits.
Is this new that it squashes the commits? In the past, I was seeing all my individual commits after the merge.

**Comment by cheshire** [comment]
No it was old, are you saying previously you saw PR being committed as multiple separate commits? I don't think that was ever the case?

**Review by cheshire — APPROVED** [review]

**Comment by nouiz** [comment]
> No it was old, are you saying previously you saw PR being committed as multiple separate commits? I don't think that was ever the case?

Check this old PR of mine: https://github.com/tensorflow/tensorflow/pull/51558/

**Comment by cheshire** [comment]
seems only one commit on TF was created?

**Comment by nouiz** [comment]
I see this:

image

So one merge commit is created. But this isn't a squash. All the commits in that PR are in the log.
If they where squashed, they wouldn't be there.

Having a merge commit is normal. If the infra could include the PR description in that merge commit description, it would be useful.

Where do you see the squash? Where do you see the mangled commit description?
Could it be that public PR and internal Google contribution are handled differently related to being squashed
 or not?

**Comment by nouiz** [comment]
Any idea why it wasn't merged yet?

**Comment by cheshire** [comment]
> Any idea why it wasn't merged yet?

Great question! Let me take a look.

**Comment by cheshire** [comment]
> `ERROR: T:/src/github/tensorflow/tensorflow/python/util/BUILD:189:27: Linking tensorflow/python/util/_pywrap_checkpoint_reader.so failed: (Exit 1120): link.exe failed: error executing command`

Windows build has failed. Let me try to rerun it.

**Comment by nouiz** [comment]
The windows build was still failed.
So I rebased by PR. Maybe it will allow a clean run.
@cheshire  can you approve again?

**Review by cheshire — APPROVED** [review]

**Comment by nouiz** [comment]
The windows CI fail due to this:
```
/D_WIN32_WINNT=0x0600 /D_CRT_SECURE_NO_DEPRECATE /D_CRT_SECURE_NO_WARNINGS /D_SILENCE_STDEXT_HASH_DEPRECATION_WARNINGS /bigobj /Zm500 /J /Gy /GF /EHsc /wd4351 /wd4291 /wd4250 /wd4996 /I. /Ibazel-out/x64_windows-opt/bin /Iexternal/local_config_python /Ibazel-out/x64_windows-opt/bin/external/local_config_python /Iexternal/eigen_archive /Ibazel-out/x64_windows-op(...)): command is longer than CreateProcessW's limit (32767 characters)
```
This can't be related to my PR as it doesn't change anything in the build and doesn't changes any dependency.
Do you know how to have my PR merged?

**Comment by cheshire** [comment]
Seems like an infra issue, escalated to infra team.

**Comment by cheshire** [comment]
Infra got into a broken inconsistent state somehow.

Unfortunately, our external compiler warnings don't have -Werror, so got this:

```
ignoring return value of function declared with 'warn_unused_result' attribute [-Werror,-Wunused-result]
    module->RemoveEmbeddedComputation(old_computation);
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ~~~~~~~~~~~~~~~
```

File: tensorflow/compiler/xla/service/gpu/horizontal_loop_fusion.h
Comment (tpopp) [inline tensorflow/compiler/xla/service/gpu/horizontal_loop_fusion.h:97]: Super nit: "tenors" is a typo?

(I hope this isn't blocking. It's not important unless you happen to see it at a convenient time.)

**Comment by nouiz** [comment]
> Infra got into a broken inconsistent state somehow.
> 
> Unfortunately, our external compiler warnings don't have -Werror, so got this:
> 
> ```
> ignoring return value of function declared with 'warn_unused_result' attribute [-Werror,-Wunused-result]
>     module->RemoveEmbeddedComputation(old_computation);
>     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ~~~~~~~~~~~~~~~
> ```

Fixed.

**Review by cheshire — APPROVED** [review]

**Comment by nouiz** [comment]
@cheshire any idea why it isn't merged while the other PR got merged?

**Comment by cheshire** [comment]
Approved again internally. Sorry it's annoying multi-step process, we're trying to fix it.

**Comment by nouiz** [comment]
@cheshire Any idea why it isn't merged?

**Comment by cheshire** [comment]
Seems to be lost in copybara pipeline again. @reedwm could you investigate?

**Pull request merged**
